pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/PulsarAuthenticationException.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.authentication;
+
+import javax.naming.AuthenticationException;
+import lombok.Getter;
+
+/**
+ * PulsarAuthenticationException indicates Authentication exception in Pulsar, it contains an error code.
+ */
+public class PulsarAuthenticationException extends AuthenticationException {
+
+    public enum ErrorCode {
+        UNKNOWN,
+        PROVIDER_LIST_AUTH_REQUIRED,
+        BASIC_INVALID_TOKEN,
+        BASIC_INVALID_AUTH_DATA,
+        AUTHZ_NO_CLIENT,
+        AUTHZ_NO_TOKEN,
+        AUTHZ_NO_PUBLIC_KEY,
+        AUTHZ_DOMAIN_MISMATCH,
+        AUTHZ_INVALID_TOKEN,
+        TLS_NO_CERTS,
+        TLS_NO_CN, // cn: common name
+        TOKEN_INVALID_HEADER,
+        TOKEN_NO_AUTH_DATA,
+        TOKEN_EMPTY_TOKEN,
+        TOKEN_INVALID_TOKEN,
+        TOKEN_INVALID_AUDIENCES,
+    }
```

Comment:
My primary concern with relying on this enum is that it is not easily extended. For example, I just merged the OIDC Provider with the following enum:

https://github.com/apache/pulsar/blob/bafecb2a9c0e73942de6a38df72dd5888d5afd66/pulsar-broker-auth-oidc/src/main/java/org/apache/pulsar/broker/authentication/oidc/AuthenticationExceptionCode.java#L25-L38

From your list and from mine, my primary conclusion is that the types of errors are often provider specific, so a static enum that cannot be extended feels restrictive.

I agree that it is extremely important to limit the cardinality of these metrics, but I also think we want to make it easily extensible by users.

I think an alternative solution could work by making `PulsarAuthenticationException` abstract and letting each provider implement their own exception. Then, the provider itself could have its own enum, and when the framework calls `getErrorCode`, the implementation will convert its own enum into a string.

Reply:
Good idea! I'll change this.

Reply:
My primary concern is that `ErrorCode` has many enums that are specific to each provider and are not generally useful. What do you think about the following class?

```java
    public class PulsarAuthenticationException extends AuthenticationException {

        private final String errorCode;

        // Warning, the errorLabel must not have high cardinality
        public PulsarAuthenticationException(String message, String errorLabel) {
            super(message);
            this.errorLabel = errorLabel;
        }

        public String getErrorLabel() {
            return errorLabel;
        }

    }
```

I think it is reasonable to expect maintainers to know that the `errorLabel` must not be high cardinality, and this way, we don't have enumerations that combine unrelated labels. If a provider wants to use an enum for its own tracking, that would be perfectly reasonable, and it would probably work just like the `AuthenticationProviderOpenID`.